csvqb/csvqb/tests/unit/utils/qb/test_standardise.py

```diff
@@ -0,0 +1,171 @@
+import pandas as pd
+import pytest
+from pandas.core.arrays.categorical import Categorical
+
+from csvqb.models.cube import *
+from csvqb.utils.qb.standardise import (
+    ensure_qbcube_data_is_categorical,
+    convert_data_values_to_uri_safe_values,
+)
+
+
+def test_qbcube_data_is_categorical():
+    """
+    Ensure that when we convert a QbCube's dataframe columns to categorical, it is done for the appropriate columns.
+    """
+    data = pd.DataFrame(
+        {
+            "Some Dimension": ["a", "b", "c"],
+            "Some Resource Attribute": ["r-a", "r-b", "r-c"],
+            "Some Literal Attribute": ["l-a", "l-b", "l-c"],
+            "Measure": ["meas-a", "meas-b", "meas-c"],
+            "Unit": ["unit-a", "unit-b", "unit-c"],
+            "Observed Value": [1, 2, 3],
+        }
+    )
+    cube = Cube(
+        CatalogMetadata("Some Qube"),
+        data,
+        [
+            QbColumn(
+                "Some Dimension",
+                NewQbDimension.from_data("Some Dimension", data["Some Dimension"]),
+            ),
+            QbColumn(
+                "Some Resource Attribute",
+                NewQbAttribute.from_data(
+                    "Some Resource Attribute", data["Some Resource Attribute"]
+                ),
+            ),
+            QbColumn(
+                "Some Literal Attribute",
+                NewQbAttributeLiteral("string", "Some Literal Attribute"),
+            ),
+            QbColumn(
+                "Measure",
+                QbMultiMeasureDimension.new_measures_from_data(data["Measure"]),
+            ),
+            QbColumn("Unit", QbMultiUnits.new_units_from_data(data["Unit"])),
+            QbColumn("Observed Value", QbMultiMeasureObservationValue()),
+        ],
+    )
+
+    ensure_qbcube_data_is_categorical(cube)
+
+    map_col_to_expected_categories = {
+        "Some Dimension": {"a", "b", "c"},
+        "Some Resource Attribute": {"r-a", "r-b", "r-c"},
+        "Measure": {"meas-a", "meas-b", "meas-c"},
+        "Unit": {"unit-a", "unit-b", "unit-c"},
+    }
+    expected_non_categorical_columns = {"Observed Value", "Some Literal Attribute"}
+
+    for (column_name, expected_categories) in map_col_to_expected_categories.items():
+        values = cube.data[column_name].values
+        assert isinstance(values, Categorical)
```

Comment:
Is there a chance for the values to be an empty string or Null value? If so what would be the outcome?

Reply:
It shouldn't happen, but if it ever did, it'd cause the unit-test to fail which I think is perfectly fine - we'll know something is wrong.